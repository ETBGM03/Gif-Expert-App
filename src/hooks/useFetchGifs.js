import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGif";

export const useFecthGifs = (category) => {
  const [state, setState] = useState({
    data: [],
    loading: true,
  });

  useEffect(() => {
    setTimeout(() => {
      getGifs(category).then((gifs) =>
        setState({
          data: gifs,
          loading: false,
        })
      );
    }, 2000);
  }, [category]);

  return state;
};
