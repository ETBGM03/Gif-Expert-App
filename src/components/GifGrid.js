import { useFecthGifs } from "../hooks/useFetchGifs";
import PropTypes from "prop-types";
import GifGridItem from "./GifGridItem";

const GifGrid = ({ category }) => {
  const { data, loading } = useFecthGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {loading && (
        <p className="animate__animated animate__flash">Loading...</p>
      )}
      <div className="card-grid">
        {data?.map((gif) => (
          <GifGridItem key={gif.id} {...gif} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
export default GifGrid;
