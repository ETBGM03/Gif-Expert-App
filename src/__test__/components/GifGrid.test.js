import React from "react";
import GifGrid from "../../components/GifGrid";
import { shallow } from "enzyme";
import { useFecthGifs } from "../../hooks/useFetchGifs";
import "@testing-library/jest-dom";

/**
 * jest.mock lo que nos va a permitir es simular las llamadas a las funciones
 * que estén en ese archivo. Lo cual podemos controlar la informacion que es retornada
 * pero ojo de forma simulada
 */
jest.mock("../../hooks/useFetchGifs");

describe("Pruebas en <GifGrid />", () => {
  let category = "Colombia";
  test("debe mostrar el componente correctamente", () => {
    useFecthGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrar item cuando se cargan imágenes useFecthGifs", () => {
    const gifs = [
      {
        id: 123,
        title: "stef",
        url: "http:localhost:3000/stef",
      },
    ];
    useFecthGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });
    const wrapper = shallow(<GifGrid category={category} />);
    /**
     * evaluar un mock con el metodo toMatchSnapshot no es tan buena idea
     * ya que lo está comparando con el html. y no es lo que queremos, sino testear
     * de una manera mejor.
     */
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
