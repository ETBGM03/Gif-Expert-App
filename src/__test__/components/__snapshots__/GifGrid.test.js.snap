// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> debe de mostrar item cuando se cargan im√°genes useFecthGifs 1`] = `
<Fragment>
  <h3>
    Colombia
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id={123}
      key="123"
      title="stef"
      url="http:localhost:3000/stef"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en <GifGrid /> debe mostrar el componente correctamente 1`] = `
<Fragment>
  <h3>
    Colombia
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
